### PR TITLE
Fixing bug #915

### PR DESCRIPTION
**Pull request by msmith491**
Wildcard ignores cause ag to discard all result files.  This commit
addresses this issue by first checking if a wildcard '*' gitignore is
present then performs a negative check against all files that would be
disqualified by the wildcard ignore with any negative-ignores
(represented by entries starting with '!' in a .gitignore file).  If a
negative match is found, the previous match against '*' is skipped.

### CODE REVIEW

**Comment by bmalehorn** [comment]
Neat commit. Certainly we all want more .gitignore to be more accurate.

But, this commit only targets a certain case, the wildcard ignore. It doesn't fix the underlying problem.

Say you have the file `file.txt`. This change will still fail to select `file.txt` with this `.gitignore`:

    file*
    !*.txt

Also, it will incorrect select file.txt with this `.gitignore`:

    *
    !*.txt
    *

Really the change should happen in `path_ignore_search()`, iterating through each regex and ignoring and un-ignoring the file as you go along (no "early exit" as you've programmed in).

But I just tried that and it failed because we sort `.gitignores` as part of an optimization. So the above `.gitignore` will be turned into

    !*.txt
    *
    *

which is *not* the same thing. So we'd have to throw out that sorting code, or modify it quite a bit to make sure it's still accurate in the face of un-ignores. Which is probably why un-ignores are not supported at the moment.

tl;dr to really fix the problem, rather than hack this one case, you need to do a lot of work. But you're welcome to give it a try :)

**Comment by ssbarnea** [comment]
I cannot believe I spend almost an hour trying to find workarounds for this bug before finding a 3 year old PR that is fixing it. Is this project dead?
